dm/pkg/checker/table_structure.go

```diff
@@ -60,87 +74,121 @@ func (o *incompatibilityOption) String() string {
 	)
-	for schema, tables := range c.tables {
-		if len(tables) == 0 {
-			tables, err = dbutil.GetTables(ctx, c.db, schema)
-			if err != nil {
-				markCheckError(r, err)
-				return r
-			}
-		}
-
-		for _, table := range tables {
-			tableName := dbutil.TableName(schema, table)
-			statement, err := dbutil.GetCreateTableSQL(ctx, c.db, schema, table)
-			if err != nil {
-				// continue if table was deleted when checking
-				if isMySQLError(err, mysql.ErrNoSuchTable) {
-					continue
+	defer close(inCh)
+
+	checkCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	log.L().Logger.Info("start to check tables", zap.Int("concurrency", concurrency), zap.Int("all table num", len(c.tables)))
+	checkFunc := func() {
+		defer checkWg.Done()
+		for {
+			select {
+			case <-checkCtx.Done():
+				return
+			case table, ok := <-inCh:
+				if !ok {
+					return
+				}
+				statement, err := dbutil.GetCreateTableSQL(ctx, c.db, table.Schema, table.Name)
+				if err != nil {
+					// continue if table was deleted when checking
+					if isMySQLError(err, mysql.ErrNoSuchTable) {
+						continue
+					}
+					reMu.Lock()
+					markCheckError(r, err)
+					reMu.Unlock()
+					cancel()
+					break
 				}
-				markCheckError(r, err)
-				return r
-			}
 
-			opts := c.checkCreateSQL(ctx, statement)
-			if len(opts) > 0 {
-				options[tableName] = opts
-				statements[tableName] = statement
+				opts := c.checkCreateSQL(ctx, statement)
+				for _, opt := range opts {
+					opt.tableID = table.String()
+					tableMsg := "table " + opt.tableID + " "
+					reMu.Lock()
+					switch opt.state {
+					case StateWarning:
+						if r.State != StateFailure {
+							r.State = StateWarning
+						}
+						e := NewError(tableMsg + opt.errMessage)
+						e.Severity = StateWarning
+						e.Instruction = opt.instruction
+						r.Errors = append(r.Errors, e)
+					case StateFailure:
+						r.State = StateFailure
+						e := NewError(tableMsg + opt.errMessage)
+						e.Instruction = opt.instruction
+						r.Errors = append(r.Errors, e)
+					}
+					reMu.Unlock()
+				}
 			}
 		}
 	}
 
-	for name, opts := range options {
-		if len(opts) == 0 {
-			continue
-		}
-		tableMsg := "table " + name + " "
+	for i := 0; i < concurrency; i++ {
+		checkWg.Add(1)
+		go checkFunc()
+	}
 
-		for _, option := range opts {
-			switch option.state {
-			case StateWarning:
-				if len(r.State) == 0 {
-					r.State = StateWarning
-				}
-				e := NewError(tableMsg + option.errMessage)
-				e.Severity = StateWarning
-				e.Instruction = option.instruction
-				r.Errors = append(r.Errors, e)
-			case StateFailure:
-				r.State = StateFailure
-				e := NewError(tableMsg + option.errMessage)
-				e.Instruction = option.instruction
-				r.Errors = append(r.Errors, e)
-			}
+	defer func() {
+		checkWg.Wait()
+		log.L().Logger.Info("check table structure over", zap.Time("start time", startTime), zap.String("speed time", time.Since(startTime).String()))
+	}()
+
+	for _, table := range c.tables {
```

Comment:
```suggestion
	for _, table := range c.tables {
	     inCh <- table
	}
```
no need to  use select here because i think enqueue is very fast because `inCh    = make(chan *filter.Table, len(c.tables))`